controllers/goalertintegration/goalertintegration_controller.go

```diff
@@ -0,0 +1,139 @@
+)
+
+const (
+	GoalertFinalizerPrefix = "goalert.managed.openshift.io/goalert-"
+)
+
+// GoalertIntegrationReconciler reconciles a GoalertIntegration object
+type GoalertIntegrationReconciler struct {
+	client.Client
+	Scheme    *runtime.Scheme
+	reqLogger logr.Logger
+}
+
+//+kubebuilder:rbac:groups=goalert.managed.openshift.io,resources=goalertintegrations,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=goalert.managed.openshift.io,resources=goalertintegrations/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=goalert.managed.openshift.io,resources=goalertintegrations/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the GoalertIntegration object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile
+func (r *GoalertIntegrationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = log.FromContext(ctx)
+
+	// TODO(user): your logic here
+	gi := &goalertv1alpha1.GoalertIntegration{}
+	err := r.Get(context.TODO(), req.NamespacedName, gi)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return r.doNotRequeue()
+		}
+		// Error reading the object - requeue the request.
+		return r.requeueOnErr(err)
+	}
+
+	// fetch all CDs so we can inspect if they're dropped out of the matching CD list
+	allClusterDeployments, err := r.getAllClusterDeployments()
+	if err != nil {
+		return r.requeueOnErr(err)
+	}
+
+	// Fetch ClusterDeployments matching the GI's ClusterDeployment label selector
+	matchingClusterDeployments, err := r.getMatchingClusterDeployments(gi)
+	if err != nil {
+		return r.requeueOnErr(err)
+	}
+
+	// Add authentication to GraphQL API OSD-16252
+	if err != nil {
+		r.reqLogger.Error(err, "Failed to load Goalert API key from Secret listed in GoalertIntegration CR")
+	}
+
+	goalertFinalizer := GoalertFinalizerPrefix + gi.Name
+	//If the GI is being deleted, clean up all ClusterDeployments with matching finalizers
+	if gi.DeletionTimestamp != nil {
+		for i := range matchingClusterDeployments.Items {
+			clusterdeployment := allClusterDeployments.Items[i]
+			if utils.HasFinalizer(&clusterdeployment, goalertFinalizer) {
+				// Handle deletion of cluster OSD-16305
+			}
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *GoalertIntegrationReconciler) getAllClusterDeployments() (*hivev1.ClusterDeploymentList, error) {
```

Comment:
```suggestion
func (r *GoalertIntegrationReconciler) getAllClusterDeployments(ctx context.Context) (*hivev1.ClusterDeploymentList, error) {
```

Should propagate contexts through function calls